core/src/repair/quic_endpoint.rs

```diff
@@ -0,0 +1,624 @@
+            .map(ByteBuf::into_vec)
+            .ok()
+    })
+    .while_some()
+    .try_for_each(|chunk| {
+        response_sender
+            .send((remote_address, chunk))
+            .map_err(|err| {
+                close_quic_endpoint(endpoint);
+                Error::from(err)
+            })
+    })
+}
+
+async fn get_connection(
+    endpoint: &Endpoint,
+    remote_address: SocketAddr,
+    remote_request_sender: Sender<RemoteRequest>,
+    cache: Arc<RwLock<ConnectionCache>>,
+) -> Result<Connection, Error> {
+    let entry = get_cache_entry(remote_address, &cache).await;
+    {
+        let connection: Option<Connection> = entry.read().await.clone();
+        if let Some(connection) = connection {
+            if connection.close_reason().is_none() {
+                return Ok(connection);
+            }
+        }
+    }
+    let connection = {
+        // Need to write lock here so that only one task initiates
+        // a new connection to the same remote_address.
+        let mut entry = entry.write().await;
+        if let Some(connection) = entry.deref() {
+            if connection.close_reason().is_none() {
+                return Ok(connection.clone());
+            }
+        }
+        let connection = endpoint
+            .connect(remote_address, CONNECT_SERVER_NAME)?
+            .await?;
+        entry.insert(connection).clone()
+    };
+    tokio::task::spawn(handle_connection_error(
+        endpoint.clone(),
+        connection.remote_address(),
+        get_remote_pubkey(&connection)?,
+        connection.clone(),
+        remote_request_sender,
+        cache,
+    ));
+    Ok(connection)
+}
+
+fn get_remote_pubkey(connection: &Connection) -> Result<Pubkey, Error> {
+    match solana_streamer::nonblocking::quic::get_remote_pubkey(connection) {
+        Some(remote_pubkey) => Ok(remote_pubkey),
+        None => {
+            connection.close(
+                CONNECTION_CLOSE_ERROR_CODE_INVALID_IDENTITY,
+                CONNECTION_CLOSE_REASON_INVALID_IDENTITY,
+            );
+            Err(Error::InvalidIdentity(connection.remote_address()))
+        }
+    }
+}
+
+async fn get_cache_entry(
+    remote_address: SocketAddr,
+    cache: &RwLock<ConnectionCache>,
+) -> Arc<RwLock<Option<Connection>>> {
+    let key = (remote_address, /*remote_pubkey:*/ None);
+    if let Some(entry) = cache.read().await.get(&key) {
+        return entry.clone();
+    }
+    cache.write().await.entry(key).or_default().clone()
+}
+
+async fn cache_connection(
+    remote_address: SocketAddr,
+    remote_pubkey: Pubkey,
+    connection: Connection,
+    cache: &RwLock<ConnectionCache>,
+) {
+    let entries: [Arc<RwLock<Option<Connection>>>; 2] = {
+        let mut cache = cache.write().await;
+        [Some(remote_pubkey), None].map(|remote_pubkey| {
+            let key = (remote_address, remote_pubkey);
+            cache.entry(key).or_default().clone()
+        })
+    };
+    let mut entry = entries[0].write().await;
```

Comment:
Not clear to me what is the second entry with null pubkey doing, can you document?